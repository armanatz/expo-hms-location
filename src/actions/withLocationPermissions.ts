import { ExpoConfig } from 'expo/config';
import { AndroidConfig } from 'expo/config-plugins';

const { withPermissions } = AndroidConfig.Permissions;

export default function withLocationPermissions(
  config: ExpoConfig,
  isBackgroundLocationEnabled?: boolean,
) {
  return withPermissions(
    config,
    [
      'android.permission.ACCESS_COARSE_LOCATION',
      'android.permission.ACCESS_FINE_LOCATION',
      isBackgroundLocationEnabled &&
        'android.permission.ACCESS_BACKGROUND_LOCATION',
    ].filter(Boolean) as string[],
  );
}
