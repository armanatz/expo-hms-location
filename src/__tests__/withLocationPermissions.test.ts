import { ExpoConfig } from 'expo/config';

import withLocationPermissions from '../actions/withLocationPermissions';

describe('expo config android.permissions modifications', () => {
  it('adds location permissions', () => {
    const sampleExpoConfig: ExpoConfig = {
      slug: 'testproject',
      version: '1',
      name: 'testproject',
      platforms: ['ios', 'android'],
    };

    // Test that it adds permissions correctly
    const firstResult = withLocationPermissions(sampleExpoConfig);
    expect(firstResult).toMatchSnapshot();

    // Test that it doesn't duplicate permissions
    const secondResult = withLocationPermissions(sampleExpoConfig);
    expect(secondResult).toBe(firstResult);
  });

  it('adds background location permissions', () => {
    const sampleExpoConfig: ExpoConfig = {
      slug: 'testproject',
      version: '1',
      name: 'testproject',
      platforms: ['ios', 'android'],
    };

    // Test that it adds permissions - including background location correctly
    const firstResult = withLocationPermissions(sampleExpoConfig, true);
    expect(firstResult).toMatchSnapshot();

    // Test that it doesn't duplicate permissions
    const secondResult = withLocationPermissions(sampleExpoConfig, true);
    expect(secondResult).toBe(firstResult);
  });
});
