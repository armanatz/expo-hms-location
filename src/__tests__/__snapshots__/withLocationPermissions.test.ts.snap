// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expo config android.permissions modifications adds background location permissions 1`] = `
{
  "android": {
    "permissions": [
      "android.permission.ACCESS_COARSE_LOCATION",
      "android.permission.ACCESS_FINE_LOCATION",
      "android.permission.ACCESS_BACKGROUND_LOCATION",
    ],
  },
  "mods": {
    "android": {
      "manifest": [Function],
    },
  },
  "name": "testproject",
  "platforms": [
    "ios",
    "android",
  ],
  "slug": "testproject",
  "version": "1",
}
`;

exports[`expo config android.permissions modifications adds location permissions 1`] = `
{
  "android": {
    "permissions": [
      "android.permission.ACCESS_COARSE_LOCATION",
      "android.permission.ACCESS_FINE_LOCATION",
    ],
  },
  "mods": {
    "android": {
      "manifest": [Function],
    },
  },
  "name": "testproject",
  "platforms": [
    "ios",
    "android",
  ],
  "slug": "testproject",
  "version": "1",
}
`;
